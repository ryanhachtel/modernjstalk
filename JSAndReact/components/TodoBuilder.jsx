import React, { Component } from 'react';
import { Input, Button, Form, Container, Grid, Icon } from 'semantic-ui-react'
import TodoList from './TodoList';

class TodoBuilder extends Component {
	constructor(props) {
		super(props);
		
		// Defining the starting app state
		this.state = {
			todoList: [
				{ 'info': 'Completed Application', 'isCompleted': false, id: 1 },
				{ 'info': 'Win a million dollars', 'isCompleted': false, id: 2 }
			],
			filter: 'all',
			info: '',
			counter: 3
		};

		// Binding the action functions to this, so they have the context they require
		this.addTodo = this.addTodo.bind(this);
		this.updateFilter = this.updateFilter.bind(this);
		this.handleInfoChange = this.handleInfoChange.bind(this);
		this.markTodoComplete = this.markTodoComplete.bind(this);
		this.markTodoUncomplete = this.markTodoUncomplete.bind(this);
		this.deleteTodo = this.deleteTodo.bind(this);
		this.removeAllCompleted = this.removeAllCompleted.bind(this);
	}

	// Updates info state as the user types in their todo
	handleInfoChange(event) {
		this.setState({ info: event.target.value });
	}

	// Adds a new todo to the todoList
	addTodo() {
		event.preventDefault();
		const { todoList, counter } = this.state;

		const newTodo = {
			'info': this.state.info,
			'isCompleted': false,
			'id': counter
		};
		// Update the state
		this.setState({
			todoList: [...todoList, newTodo ],
			counter: counter + 1
		});
		// Reset info to empty string
		this.setState({
			info: ''
		});
	}

	// Sets the filter value depending on what filter button was clicked
	updateFilter(filter) {
		this.setState({ filter: filter });
	}

	// Marks a specific todo as complete and updates the todoList
	markTodoComplete(id) {
		const { todoList } = this.state;
		todoList.forEach((item) => {
			if (item.id === id) {
				item.isCompleted = true;
			}
		});
		this.setState({ todoList: todoList });
	}

	// Marks a specific todo as uncomplete and updates the todoList
	markTodoUncomplete(id) {
		const { todoList } = this.state;
		todoList.forEach((item) => {
			if (item.id === id) {
				item.isCompleted = false;
			}
		});
		this.setState({ todoList: todoList });
	}

	// Deletes a specific todo from the todoList
	deleteTodo(id) {
		const { todoList } = this.state;
		const toRemoveIndex = todoList.findIndex(todo => todo.id === id);
		todoList.splice(toRemoveIndex, 1);

		this.setState({ todoList: todoList});
	}
	
	// Deletes all completed todos from the todoList
	removeAllCompleted() {
		const { todoList } = this.state;
		const updatedList = todoList.filter(todo => todo.isCompleted === false);
		this.setState({ todoList: updatedList });
	}

	render() {
		// Destructure state variables
		const { todoList, filter } = this.state;
		return (
			<Container>
				<Form onSubmit={this.addTodo}>
					<Form.Field>
						<Input placeholder='Todo' size='large' icon='plus' value={this.state.info} onChange={this.handleInfoChange} />
					</Form.Field>
					<Button type='submit'>Submit</Button>
				</Form>
				<br />
				<Container>
					<Grid divided='vertically'>
						<Grid.Row columns={3}>
							<Grid.Column textAlign='center'>
								<Button size='large' icon labelPosition='left' onClick={() => this.updateFilter('completed')}>
									Show Only Completed
									<Icon name='filter' />
								</Button>
							</Grid.Column>
							<Grid.Column textAlign='center'>
								<Button size='large' icon labelPosition='left' onClick={() => this.updateFilter('uncompleted')}>
									Show Only Uncompleted
									<Icon name='filter' />
								</Button>
							</Grid.Column>
							<Grid.Column textAlign='center'>
								<Button size='large' icon labelPosition='left' onClick={() => this.updateFilter('all')}>
									Show All
									<Icon name='list' />
								</Button>
							</Grid.Column>
						</Grid.Row>
					</Grid>
					<br />
					<TodoList
						todoList={todoList}
						filter={filter}
						markTodoComplete={this.markTodoComplete}
						markTodoUncomplete={this.markTodoUncomplete}
						deleteTodo={this.deleteTodo}
					/>
				</Container>
				<br />
				<br />
				<Grid divided='vertically'>
					<Grid.Row columns={1}>
						<Grid.Column textAlign='center'>
							<Button size='large' icon labelPosition='left' onClick={() => this.removeAllCompleted()}>
								Remove All Completed
								<Icon name='remove circle' />
							</Button>
						</Grid.Column>
					</Grid.Row>
				</Grid>
			</Container>
		);
	}
}

export default TodoBuilder;
