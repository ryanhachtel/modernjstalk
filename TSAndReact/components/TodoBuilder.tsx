import * as React from 'react';
import { Input, Button, Form, Container, Grid, Icon } from 'semantic-ui-react'
import TodoList from './TodoList';
import { ITodo, ITodoBuilderProps, ITodoBuilderState, Filters } from '../types/types';

class TodoBuilder extends React.Component<ITodoBuilderProps, ITodoBuilderState>{
	constructor(props: ITodoBuilderProps) {
		super(props);
		// Defining the starting app state
		this.state = {
			todoList: [
				{ 'info': 'Completed Application', 'isCompleted': false, id: 1 },
				{ 'info': 'Win a million dollars', 'isCompleted': false, id: 2 }
			],
			filter: Filters.all,
			info: '',
			counter: 3
		};
	}

	// Updates info state as the user types in their todo
	handleInfoChange = (event: React.BaseSyntheticEvent) => {
		this.setState({ info: event.target.value });
	}

	// Adds a new todo to the todoList
	addTodo = () => {
		const { todoList, counter } = this.state;

		const newTodo: ITodo = {
			'info': this.state.info,
			'isCompleted': false,
			'id': counter
		};

		// Update the state
		this.setState({
			todoList: [...todoList, newTodo ],
			counter: counter + 1
		});

		// Reset info to empty string
		this.setState({
			info: ''
		});
	}

	// Sets the filter value depending on what filter button was clicked
	updateFilter = (filter: Filters) => {
		this.setState({ filter: filter });
	}

	// Marks a specific todo as complete and updates the todoList
	markTodoComplete = (id: number) => {
		const { todoList } = this.state;
		todoList.forEach((item) => {
			if (item.id === id) {
				item.isCompleted = true;
			}
		});
		this.setState({ todoList: todoList });
	}

	// Marks a specific todo as uncomplete and updates the todoList
	markTodoUncomplete = (id: number) => {
		const { todoList } = this.state;
		todoList.forEach((item: ITodo) => {
			if (item.id === id) {
				item.isCompleted = false;
			}
		});
		this.setState({ todoList: todoList });
	}

	// Deletes a specific todo from the todoList
	deleteTodo = (id: number) => {
		const { todoList } = this.state;
		const toRemoveIndex: number = todoList.findIndex(todo => todo.id === id);
		todoList.splice(toRemoveIndex, 1);

		this.setState({ todoList: todoList});
	}
	
	// Deletes all completed todos from the todoList
	removeAllCompleted = () => {
		const { todoList } = this.state;
		const updatedList = todoList.filter(todo => todo.isCompleted === false);
		this.setState({ todoList: updatedList });
	}

	render() {
		// Destructure state variables
		const { todoList, filter } = this.state;
		return (
			<Container>
				<Form onSubmit={this.addTodo}>
					<Form.Field>
						<Input placeholder='Todo' aria-label='addNewTodoText' size='large' icon='plus' value={this.state.info} onChange={this.handleInfoChange} />
					</Form.Field>
					<Button type='submit' data-testid='addNewTodoSubmit'>Submit</Button>
				</Form>
				<br />
				<Container>
					<Grid divided='vertically'>
						<Grid.Row columns={3}>
							<Grid.Column textAlign='center'>
								<Button size='large' icon labelPosition='left' onClick={() => this.updateFilter(Filters.completed)}>
									Show Only Completed
									<Icon name='filter' />
								</Button>
							</Grid.Column>
							<Grid.Column textAlign='center'>
								<Button size='large' icon labelPosition='left' onClick={() => this.updateFilter(Filters.uncompleted)}>
									Show Only Uncompleted
									<Icon name='filter' />
								</Button>
							</Grid.Column>
							<Grid.Column textAlign='center'>
								<Button size='large' icon labelPosition='left' onClick={() => this.updateFilter(Filters.all)}>
									Show All
									<Icon name='list' />
								</Button>
							</Grid.Column>
						</Grid.Row>
					</Grid>
					<br />
					<TodoList
						todoList={todoList}
						filter={filter}
						markTodoComplete={this.markTodoComplete}
						markTodoUncomplete={this.markTodoUncomplete}
						deleteTodo={this.deleteTodo}
					/>
				</Container>
				<br />
				<br />
				<Grid divided='vertically'>
					<Grid.Row columns={1}>
						<Grid.Column textAlign='center'>
							<Button size='large' icon labelPosition='left' onClick={() => this.removeAllCompleted()}>
								Remove All Completed
								<Icon name='remove circle' />
							</Button>
						</Grid.Column>
					</Grid.Row>
				</Grid>
			</Container>
		);
	}
}

export default TodoBuilder;
