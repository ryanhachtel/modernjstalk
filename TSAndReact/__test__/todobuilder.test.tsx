import * as React from 'react';
import TodoBuilder from '../components/TodoBuilder';
import * as renderer from 'react-test-renderer';

test('Renders correctly', () => {
	// @ts-ignore
	const component = renderer.create(<TodoBuilder />);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

