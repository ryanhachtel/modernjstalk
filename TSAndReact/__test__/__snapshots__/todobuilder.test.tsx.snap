// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="ui container"
>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <div
        className="ui large icon input"
      >
        <input
          aria-label="addNewTodoText"
          onChange={[Function]}
          placeholder="Todo"
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          className="plus icon"
          onClick={[Function]}
        />
      </div>
    </div>
    <button
      className="ui button"
      data-testid="addNewTodoSubmit"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
  <br />
  <div
    className="ui container"
  >
    <div
      className="ui vertically divided grid"
    >
      <div
        className="three column row"
      >
        <div
          className="center aligned column"
        >
          <button
            className="ui large icon left labeled button"
            onClick={[Function]}
          >
            Show Only Completed
            <i
              aria-hidden="true"
              className="filter icon"
              onClick={[Function]}
            />
          </button>
        </div>
        <div
          className="center aligned column"
        >
          <button
            className="ui large icon left labeled button"
            onClick={[Function]}
          >
            Show Only Uncompleted
            <i
              aria-hidden="true"
              className="filter icon"
              onClick={[Function]}
            />
          </button>
        </div>
        <div
          className="center aligned column"
        >
          <button
            className="ui large icon left labeled button"
            onClick={[Function]}
          >
            Show All
            <i
              aria-hidden="true"
              className="list icon"
              onClick={[Function]}
            />
          </button>
        </div>
      </div>
    </div>
    <br />
    <div>
      <div
        className="ui divided middle aligned list"
        role="list"
      >
        <div
          className="item"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="right floated content"
          >
            <button
              className="ui large icon button"
              name="check"
              onClick={[Function]}
            >
              <i
                aria-hidden="true"
                className="check circle outline icon"
                onClick={[Function]}
              />
            </button>
            <button
              className="ui large icon button"
              name="trash"
              onClick={[Function]}
            >
              <i
                aria-hidden="true"
                className="trash icon"
                onClick={[Function]}
              />
            </button>
          </div>
          <i
            aria-hidden="true"
            className="pointing right large icon"
            onClick={[Function]}
          />
          <div
            className="content"
          >
            <h3
              className="header uncompleted"
            >
              Completed Application
            </h3>
          </div>
        </div>
        <div
          className="item"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="right floated content"
          >
            <button
              className="ui large icon button"
              name="check"
              onClick={[Function]}
            >
              <i
                aria-hidden="true"
                className="check circle outline icon"
                onClick={[Function]}
              />
            </button>
            <button
              className="ui large icon button"
              name="trash"
              onClick={[Function]}
            >
              <i
                aria-hidden="true"
                className="trash icon"
                onClick={[Function]}
              />
            </button>
          </div>
          <i
            aria-hidden="true"
            className="pointing right large icon"
            onClick={[Function]}
          />
          <div
            className="content"
          >
            <h3
              className="header uncompleted"
            >
              Win a million dollars
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>
  <br />
  <br />
  <div
    className="ui vertically divided grid"
  >
    <div
      className="one column row"
    >
      <div
        className="center aligned column"
      >
        <button
          className="ui large icon left labeled button"
          onClick={[Function]}
        >
          Remove All Completed
          <i
            aria-hidden="true"
            className="remove circle icon"
            onClick={[Function]}
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
